src/dbnode/storage/bootstrap/types.go

```diff
@@ -416,5 +420,51 @@ type Source interface {
 	// A bootstrapper source should only return an error should it want to
 	// entirely cancel the bootstrapping of the node, i.e. non-recoverable
 	// situation like not being able to read from the filesystem.
-	Read(ctx context.Context, namespaces Namespaces) (NamespaceResults, error)
+	Read(ctx context.Context, namespaces Namespaces, state State) (NamespaceResults, error)
+}
+
+// InfoFileResultsPerShard maps shards to info files.
+type InfoFileResultsPerShard map[uint32][]fs.ReadInfoFileResult
+
+// InfoFilesByNamespace maps a namespace to info files grouped by shard.
+type InfoFilesByNamespace map[namespace.Metadata]InfoFileResultsPerShard
+
+// State provides a snapshot of info files for use throughout all stages of the bootstrap.
+type State struct {
+	fsOpts               fs.Options
+	infoFilesFinders     []InfoFilesFinder
+	infoFilesByNamespace InfoFilesByNamespace
+	iOpts                instrument.Options
+}
+
+// StateOptions represents the options for State.
+type StateOptions interface {
+	// Validate will validate the options and return an error if not valid.
+	Validate() error
+
+	// SetFilesystemOptions sets the filesystem options.
+	SetFilesystemOptions(value fs.Options) StateOptions
+
+	// FilesystemOptions returns the filesystem options.
+	FilesystemOptions() fs.Options
+
+	// SetInfoFilesFinders sets the finders used to load info files for all namespaces provided.
+	SetInfoFilesFinders(value []InfoFilesFinder) StateOptions
```

Comment:
nit: see comment below but maybe we could use something generic that already exists for bootstrapping like the `bootstrap.Namespace` struct and then these methods could be more generically named like so.
```
SetNamespaces(namespaces []Namespace) StateOptions
Namespaces() []Namespace
```